backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractService.java

```diff
@@ -399,4 +406,73 @@ public void cancelContractsOnOrderVoidIfNeededAndAllowed(@NonNull final OrderId
 													.build());
 		}
 	}
+
+	public void closeContractsOnOrderCloseIfNeededAndAllowed(@NonNull final I_C_Order orderRecord)
+	{
+		final List<de.metas.interfaces.I_C_OrderLine> orderLines = orderBL.retrieveOrderLines(orderRecord);
+		final Set<FlatrateTermId> modularContractIds = orderLines.stream()
+				.filter(this::isModularContractLine)
+				.map(I_C_OrderLine::getC_Order_ID)
+				.map(OrderId::ofRepoId)
+				.map(flatrateBL::getByOrderId)
+				.flatMap(List::stream)
+				.map(I_C_Flatrate_Term::getC_Flatrate_Term_ID)
+				.map(FlatrateTermId::ofRepoId)
+				.collect(Collectors.toSet());
+		final boolean isModularOrder = !modularContractIds.isEmpty();
+
+		if (!isModularOrder)
+		{
+			return;
+		}
+
+		final boolean hasBillableUnprocessedLogs = modularContractLogService.anyMatch(
+				ModularContractLogQuery.builder()
+						.flatrateTermIds(modularContractIds)
+						.billable(true)
+						.processed(false)
+						.build()
+		);
+
+		if(hasBillableUnprocessedLogs)
+		{
+			throw new AdempiereException("Not all billable logs are processed"); // TODO
+		}
+
+		if(!orderRecord.isSOTrx())
+		{
+			final Set<OrderId> relatedSalesOrdersIds = queryBL.createQueryBuilder(I_C_OrderLine.class)
+					.addOnlyActiveRecordsFilter()
+					.addInArrayFilter(I_C_OrderLine.COLUMNNAME_Purchase_Modular_Flatrate_Term_ID, modularContractIds)
+					.create()
+					.stream()
+					.map(I_C_OrderLine::getC_Order_ID)
+					.map(OrderId::ofRepoId)
+					.collect(Collectors.toSet());
+
+			final boolean notClosedSalesOrderExists = relatedSalesOrdersIds.stream().anyMatch(not(orderBL::isClosed));
+			if(notClosedSalesOrderExists)
+			{
+				throw new AdempiereException("All Sales Orders linked to this Orders Modular Contract should be closed first"); //TODO
+			}
+		}
+
+		modularContractIds.forEach(flatrateTermId -> contractChangeBL.endContract(flatrateBL.getById(flatrateTermId)));
+
+		if(orderRecord.isSOTrx())
+		{
+			queryBL.createQueryBuilder(I_C_Flatrate_Term.class)
+					.addOnlyActiveRecordsFilter()
+					.addInArrayFilter(I_C_Flatrate_Term.COLUMNNAME_Modular_Flatrate_Term_ID, modularContractIds)
+					.create()
+					.forEach(contractChangeBL::endContract);
```

Comment:
isn't this similar with line 460?!

```java
modularContractIds.forEach(flatrateTermId -> contractChangeBL.endContract(flatrateBL.getById(flatrateTermId)));
```

Reply:
Line 460 is closing modular contracts, this here is closing the interim contracts linked to these modular contracts

Reply:
ok, but i would suggest introducing FlatrateBL/DAO.getInterimContractsByModularContractIds(...) method and use it.
in this way
* you are avoiding calling queryBL in service but DAO/repo shall do it
* your code is self documented